node/tests/utils.rs

```diff
@@ -159,16 +160,95 @@ impl MASQNode {
+        Self::wait_for_log_at_directory(
+            pattern,
+            logfile_path.as_ref(),
+            buffer,
+            &do_with_finding,
+            limit_ms,
+        );
+    }
+
+    pub fn capture_log_at_directory(
+        pattern: &str,
+        logfile_dir: &Path,
+        buffer: &mut String,
+        groups: Vec<usize>,
+        required_number_of_captures: usize,
+        limit_ms: Option<u64>,
+    ) -> Vec<Vec<String>> {
+        let logfile_path = Self::path_to_logfile(logfile_dir);
+        let do_with_findings =
+            |buffer: &String, regex: &Regex| -> (Option<Vec<Vec<String>>>, bool) {
+                let captures = regex.captures_iter(&buffer[..]).collect::<Vec<Captures>>();
+                if captures.is_empty() {
+                    return (None, false);
+                }
+                if captures.len() < required_number_of_captures {
+                    return (None, false);
+                }
+                let structured_captures = (0..captures.len())
+                    .flat_map(|idx| {
+                        captures.get(idx).map(|capture| {
+                            groups
+                                .iter()
+                                .flat_map(|group| {
+                                    capture
+                                        .get(*group)
+                                        .map(|existing_match| existing_match.as_str().to_string())
+                                })
+                                .collect::<Vec<String>>()
+                        })
+                    })
+                    .collect::<Vec<Vec<String>>>();
+                (Some(structured_captures), true)
+            };
+        Self::wait_for_log_at_directory(
+            pattern,
+            logfile_path.as_ref(),
+            buffer,
+            &do_with_findings,
+            limit_ms,
+        )
+        .unwrap()
+    }
+
+    fn wait_for_log_at_directory<T>(
+        pattern: &str,
+        path_to_logfile: &Path,
+        buffer: &mut String,
+        do_with_output: &dyn Fn(&String, &Regex) -> (Option<T>, bool),
+        limit_ms: Option<u64>,
+    ) -> Option<T> {
         let regex = regex::Regex::new(pattern).unwrap();
         let real_limit_ms = limit_ms.unwrap_or(0xFFFFFFFF);
         let started_at = Instant::now();
-        let path_to_logfile = Self::path_to_logfile(&self.data_dir);
         loop {
             match std::fs::read_to_string(&path_to_logfile) {
                 Ok(contents) => {
-                    self.logfile_contents = contents;
-                    if regex.is_match(&self.logfile_contents[..]) {
-                        break;
+                    buffer.clear();
+                    buffer.push_str(&contents);
+                    let (result_opt, finished) = do_with_output(&buffer, &regex);
+                    if finished {
+                        break result_opt;
                     }
                 }
                 Err(e) => {
```

Comment:
If the error here stays around regardless of how long you wait, then this is a potential infinite loop. Either it should immediately fail the test, or it should check the time and fail the test if the timeout has expired.